### ISSUE TITLE
Critical errors with a program in lvl controller after loading saved game

### ISSUE THREAD

**Issue by rbcat** [rbcat/returbot#1]
This particular trouble appears in the level 4-4 "Explosives". When I start the game, save the game and then load the save, either one of two things happen:
1) Level controller stops working by throwing "illegal object" error, and that's all.
2) Level controller does not stops working. But when I kill all three spiders in the first camp northward from the space ship, thus triggering respawn, the game just crashes.

I encounter this problem by using level controller, but about two things I am certain:
1) This is not related to level controllers per se, because if I use ordinary bots instead results are just as bad.
2) It doesn't matter if you load your program to object (a level controller or a bot) in scene.txt or manually after starting the game.

Basically, I started a [topic](https://colobot.info/forum/showthread.php?tid=969) in the forum in regards to this trouble.

**Comment by melex750** [comment]
Confirmed with Colobot: Gold Edition 0.1.11.
Using gdb, the game always crashes while executing one of [these lines](https://github.com/rbcat/returbot/blob/master/returbot/levels/custom/rb04/level004/Control.txt#L249-L253).
Specifically while executing ```lim = camps.limX;```, where X is 0, 1, 2, 3, or 4.
Some data isn't being saved for that expression.

I was able to fix the level controller script so doesn't crash after loading a save and killing spiders,
and new spiders are re-spawning. Generally speaking, using setters and getters for private static
data will help avoid such issues. I can submit a pull request later if you want. 

Nice project by the way, I've had fun testing a few levels.

**Comment by rbcat** [comment]
Woooo... You're really are a hero, @melex750. Thanks a lot for that. I tested it and there indeed were no more crashes. Merged your commit.

**Comment by melex750** [comment]
@rbcat

> BTW, an issue with "illegal object" was caused by search() instruction and it was fixed before.
> Seems like calling search() in a running program within if() after loading a savegame was causing it.

I believe you are referring to **[this code here](https://github.com/rbcat/returbot/commit/1035acc5c7b91db7e48a3ac40d6469b93cce6207#diff-1740914d7bc339da238442b2f04ee594L240)**.
This likely happens if the game is saved at the dot between search() and position, in ```search().position```.

That code seems to be checking for Aliens in a certain area.
It can be simplified a lot with more search() parameters.
```search ( cat, pos, min, max, sens, filter );```
```c++
if ( null == search(list, camps.camps[i], 0.0, camps.awd[i]) )
{
    camps.awd[i] = -1;
}
```

You can count aliens in a certain area too, with ```searchall()```:
```c++
int numSpiders = sizeof( searchall(AlienSpider, pos, 0, 40) );
```

```searchall ( cat, pos, min, max, sens, filter );```

You can also catch the error with ```try...catch``` like this:
```c++
try
{
    // code that may throw an error
    search(AlienSpider).position;
}
catch( CBotErrDeletedPtr ) // "Illegal object" sometimes happens with savedgame
{
    continue;
}
catch( CBotErrNull ) // "Unknown Object" in range of search() or radar()
{ // tried to use a pointer that is null
    continue;
}
```

**Comment by rbcat** [comment]
> This likely happens if the game is saved at the dot between search() and position, in search().position.
It makes a sense, but the very same savegame was causing "illegal error" not always, only sometimes. And I wonder why doesn't it happens with radar().

I forgot that search() was extended and searchall() was added as well. I was then looking into documentation in SatCom which is obsolete, but now you made me remember. I took the 1st solution of yours.
